### PR TITLE
Reimplement dng opcodes

### LINKED ISSUE #79: Bad pixel correction for DNGs is broken
See https://redmine.darktable.org/issues/10078.

Example image: https://raw.pixls.us/getfile.php/683/raw/RAW_CANON_POWERSHOT_SX130IS_CHDK.DNG

I stumbled on this issues while cleaning up the DngOpcode classes. This has almost certainly never worked (I found 2 separate issues).

I have a complete rewrite of the DngOpcode classes, fixing this issue (among others). Just waiting for #75 to be merged.

### PR DESCRIPTION
**Pull request by axxel**
Fix #79 and a bunch of other issues (interface bloat, copy'n'paste errors, code duplication, leaks on exceptions). Details see commit messages.

### CODE REVIEW

File: src/librawspeed/common/Common.h [hunk 1]
```diff
@@ -130,21 +130,39 @@ inline Endianness getHostEndianness() {
...
-  if (bswap) {
-    switch(sizeof(T)) {
-    case 1: break;
-    case 2: ret = BSWAP16(ret); break;
-    case 4: ret = BSWAP32(ret); break;
-    case 8: ret = BSWAP64(ret); break;
-    }
-  }
-  return ret;
+  return bswap ? getByteSwapped(ret) : ret;
```
Comment (LebedevRI) [inline src/librawspeed/common/Common.h]: While there, can you add at some very least basic test for this area of code (`getByteSwapped()`) please?

Comment (axxel) [reply src/librawspeed/common/Common.h]: Well, I'd consider those trivial functions not really worth writing elaborate tests for (I admire your patience in this regard looking at your latest pr :)). But I got this wrong the first time, so I kind of proved myself wrong. I'll add something to your new `CommonTests.cpp` once it is merged.

Comment (LebedevRI) [reply src/librawspeed/common/Common.h]: Problem being, rstest is only run manually, while the tests are run by all the ci, and by OBS (with non-x86, and big-endian builds, hurray). So unless it is a test, problems are likely to stay unnoticed...

File: src/librawspeed/common/DngOpcodes.cpp [hunk 2]
```diff
@@ -1,698 +1,424 @@
...
+  }
+};
 
+using ScalePerRow = ScalePerRowOrCol<SelectY>;
+using ScalePerCol = ScalePerRowOrCol<SelectX>;
 
-RawImage& OpcodeScalePerCol::createOutput( RawImage &in )
+// ****************************************************************************
+
+DngOpcodes::DngOpcodes(TiffEntry *entry)
```
Comment (LebedevRI) [inline src/librawspeed/common/DngOpcodes.cpp:362]: I don't get it. How is it so different that it now results in more than twice the amount of stack frame size it used to use
And moving of the function does not help with diff either

Comment (axxel) [reply src/librawspeed/common/DngOpcodes.cpp:362]: This OSX + stack usage situation is slightly annoying. You seem to have been struggling with it for a while already. This particular constructor hardly changed at all. The `sizeof(ByteStream)` is neglectable. Since there are only very select places where the mac-code differs at all: why not completely disable the error on mac and be happy as long as the linux builds don't blow a reasonable stack size limit?

Comment (LebedevRI) [reply src/librawspeed/common/DngOpcodes.cpp:362]: 2cd77a573316143b72d5ed0b2845b8002b50b807

Comment (axxel) [reply src/librawspeed/common/DngOpcodes.cpp:362]: works, thanks.

File: src/librawspeed/common/Common.h (hunk 1 above)
Comment (LebedevRI) [reply src/librawspeed/common/Common.h]: Looking at this in more detail, `common.h` is still kinda bloaty.
i'll pick the first commit, and then move this endianness stuff into new header

Comment (LebedevRI) [reply src/librawspeed/common/Common.h]: @axxel ok, endianness stuff is in, with full tests.
please rebase this pr, and drop that commit from it
Hopefully will merge at least one of your pr's tomorrow...

Comment (axxel) [reply src/librawspeed/common/Common.h]: Two points of constructive feedback:
 * If you had simply merged that pr as is and did your split of Common.h afterwards, it would have cost you the same amount of work but spared me about 5 minutes for the manual rebase.
 * Your full tests just increased the runtime of the tests by factor of nearly 10. This is out of proportion. If you checked that e.g. `0x0102 == swap(0x0201)` then there is no point at all in also testing `0x0103`. The whole set of swap functions can be tested with full coverage in a fraction of a second (instead of somewhere near 25s on my machine).

Comment (LebedevRI) [reply src/librawspeed/common/Common.h]: > Your full tests just increased the runtime of the tests by factor of nearly 10. This is out of proportion.

I had to pick /some/ count of the samples, it happened to be the current values. Maybe there should be less samples. The actual ~problem~, is that i want more granulated output from `ctest`. Right now i switched back to less granulated output, and it should be **much** faster.

Also, you do know that you can do `ctest -j9`?

> If you had simply merged that pr as is and did your split of Common.h afterwards, it would have cost you the same amount of work but spared me about 5 minutes for the manual rebase.

Sorry, i agreed to assume maintaintership with thought of *me* being able to practice C++, and improve the library, with *everything* else being secondary. So yes, i guess this is me being somewhat discouraging from active contribution, but i also see no alternatives. Either it works for me, or it does not... Just putting all the cards on the table here.

TLDR: contributing back is important, even though the license does not require it. But *I* need to be motivated to review&merge things.

Comment (axxel) [reply src/librawspeed/common/Common.h]: > and it should be much faster.

It is indeed. thanks.

> Also, you do know that you can do ctest -j9?

No I did not. It is not that I really cared about the 30 seconds test run. I mainly wanted to point out that your testing methodology could be improved. I think a white-box testing approach is the right tool for this kind of problem and the complete set of those byte swap functions has on the order of 10 code paths (did not count). What I am ultimately trying to say is: in my personal opinion you could have spent your time and effort in this particular case to produce something with more value for the project than those bash-scripted black-box test numbers.

> Just putting all the cards on the table here.

Thanks for that. Appreciated. Just my 2 cents, though: as a respectful maintainer, maybe the good of the project could use a little bump in your priority list ;).

> But I need to be motivated to review&merge things.

Agreed. I did this hacking marathon for fun and I don't expect others to do stuff they don't like. But reviewing PRs is part of the responsibility of the role _you gave yourself_. If you don't like that just do something else. If that means not looking into code from others anymore, so be it. But if your sense of 'ownership' for the code (most of which you did not write) means forking this project at this point in time for this reason would be just sad (IMHO).

Anyway, I said I was about done with my enthusiasm working on this code and that is still true. I simply wanted to leave something 'consistent' and therefore did the Tiff interface, FILEMAP_MARGIN and DngOpcodes pr. The `Hint` class was just a little thing that seemed so obvious, I could not resist. Same goes for the Coverity fixes. I was curious if they would make sense now that you changed the compiler and while going through them one by one, why not fix them right away...

The only thing left in my patch queue at the moment is the introduction of some 'size/offset' type into the BitStream and TiffEntry interfaces. I was still waiting on some final response from you on the naming question in #76 (which really is only about "which string is nicer to look at: 'u16' or 'uint16_t'?")

TLDR; I'll let you hack on in peace soon ;).

**Comment by LebedevRI** [comment]
Is this just a rewrite, or does this fix any bugs?
In other words, how do i know that it actually works?

**Comment by axxel** [comment]
> Is this just a rewrite, or does this fix any bugs?

Please read the PR summary.

> In other words, how do i know that it actually works?

In #79, I mentioned a dt issue and gave an example file showing the problem. That file shows black pixels when rendered before the fix. Those are gone with this fix. There were other issues in the old code (like the clampBits one) that were undiscovered probably because no one used them. Maybe more issues show up later when samples are available that actually use all the implemented opcodes.
